`Composite`: use internal context to consume composite store

ciampo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Use React context to make sure that `Composite` sub-components use the correct composite context

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Since multiple ariakit components use composite stores in their internal implementation, sometimes sub-components can consume the wrong composite store (see https://github.com/WordPress/gutenberg/pull/63569#discussion_r1713930889)

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Add a new context that holds a reference to the correct composite `store`. Subcomponents get the `store` from context instead of relying on Ariakit internals

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

- Make sure that all composite instances keep working as expected in Storybook and across the editor
- Apply changes from #63569 and make sure that the regression described in https://github.com/WordPress/gutenberg/pull/63569#discussion_r1713359138 is fixed

ciampo committed:
Create CompositeContext

ciampo committed:
Use context to pass store to composite subcomponents

ciampo committed:
CHANGELOG

ciampo committed:
Remove double dot

tyxla commented on packages/components/src/composite/index.tsx:
```diff
@@ -136,9 +179,28 @@ export const Composite = Object.assign(
+	) {
+		const contextValue = useMemo(
+			() => ( {
+				store,
```
Let's document this so folks know that if they want to access the store inside a `Composite`, they can do it with the context.

ciampo replied:
Added JSDocs and README section

tyxla reviewed (approved):
LGTM 👍 🚀 

I'd only add some documentation that includes that the store is passed down through context and should be consumed through `useCompositeContext()`. 

Also, let's make sure to update the instances that still provide store explicitly to `Composite` subcomponents, for example:

https://github.com/WordPress/gutenberg/blob/2c62c6aa2f30060b3acd40b225aebe4295ad0bf4/packages/block-directory/src/components/downloadable-block-list-item/index.js#L117

and

https://github.com/WordPress/gutenberg/blob/22e4cc2e24e89901d7b0b8099a7ecc4f695e04e2/packages/dataviews/src/dataviews-layouts/list/index.tsx#L148

ciampo committed:
Export `Context`, add basic docs

ciampo committed:
Update CHANGELOG

ciampo commented:
> I'd only add some documentation that includes that the store is passed down through context and should be consumed through useCompositeContext().

I've added `Context` as an exported sub-component of `Composite`, and added relative JSDocs and README section. I won't export `useCompositeContext` for now because consumers can simply use `React.useContext( Composite.Context )`. We can always add the custom hook if we think it's necessary.

> Also, let's make sure to update the instances that still provide store explicitly to Composite subcomponents, for example:

Passing `store` explicitly is indeed redundant after this PR. The changes that you suggest will be applied in #63569
